File: Difficulty: Medium/Number of Ways to Arrive at Destination/number-of-ways-to-arrive-at-destination.cpp
class Solution {
public:
    int countPaths(int V, vector<vector<int>>& edges) {
        const long long MOD = 1e9 + 7;
        vector<vector<pair<int,int>>> adj(V);
        for (auto &e : edges) {
            int u = e[0], v = e[1], w = e[2];
            adj[u].push_back({v, w});
            adj[v].push_back({u, w});
        }
        vector<long long> dist(V, LLONG_MAX);
        vector<long long> ways(V, 0);
        priority_queue<
            pair<long long,int>,
            vector<pair<long long,int>>,
            greater<pair<long long,int>>
        > pq;
        dist[0] = 0;
        ways[0] = 1;
        pq.push({0, 0});
        while (!pq.empty()) {
            auto [d, u] = pq.top();
            pq.pop();
            if (d > dist[u]) continue;
            for (auto &p : adj[u]) {
                int v = p.first;
                long long w = p.second;
                if (dist[v] > dist[u] + w) {
                    dist[v] = dist[u] + w;
                    ways[v] = ways[u];
                    pq.push({dist[v], v});
                }
                else if (dist[v] == dist[u] + w) {
                    ways[v] = (ways[v] + ways[u]) % MOD;
                }
            }
        }
        return ways[V - 1] % MOD;
    }
};

File: Difficulty: Medium/Number of Ways to Arrive at Destination/number-of-ways-to-arrive-at-destination.py
import heapq

class Solution:
    def countPaths(self, V, edges):
        # Build adjacency list
        adj = [[] for _ in range(V)]
        for u, v, w in edges:
            adj[u].append((v, w))
            adj[v].append((u, w))
        
        # Min heap
        pq = []
        heapq.heappush(pq, (0, 0))  # (distance, node)
        
        # Distance and ways arrays
        dist = [float('inf')] * V
        ways = [0] * V
        
        dist[0] = 0
        ways[0] = 1
        
        while pq:
            curr_dist, node = heapq.heappop(pq)
            
            if curr_dist > dist[node]:
                continue
            
            for neighbor, weight in adj[node]:
                new_dist = curr_dist + weight
                
                # Found shorter path
                if new_dist < dist[neighbor]:
                    dist[neighbor] = new_dist
                    ways[neighbor] = ways[node]
                    heapq.heappush(pq, (new_dist, neighbor))
                
                # Found another shortest path
                elif new_dist == dist[neighbor]:
                    ways[neighbor] += ways[node]
        
        return ways[V - 1]